### PR TITLE
Gzip snapshots to db

### PR DESCRIPTION
**Pull request by petchells**

### CODE REVIEW

File: whois-nrtm4/src/main/java/net/ripe/db/nrtm4/SnapshotFileGenerator.java [hunk 1]
```diff
@@ -3,92 +3,148 @@
...
+                LOGGER.info("NRTM writer thread interrupted", e);
+                Thread.currentThread().interrupt();
             }
         }
-        return snapshotFiles;
+        LOGGER.info("NRTM generation complete {}", stopwatchRoot);
+        return publishedFiles;
     }
 
     public Optional<SnapshotFile> getLastSnapshot(final NrtmSourceModel source) {
```
Comment (MiguelAHM) [inline whois-nrtm4/src/main/java/net/ripe/db/nrtm4/SnapshotFileGenerator.java:150]: We are not using it

File: whois-nrtm4/src/main/java/net/ripe/db/nrtm4/SnapshotFileGenerator.java [hunk 2]
```diff
@@ -3,92 +3,148 @@
...
         this.snapshotFileRepository = snapshotFileRepository;
-        this.snapshotObjectSynchronizer = snapshotObjectSynchronizer;
+        this.snapshotFileSerializer = snapshotFileSerializer;
         this.sourceRepository = sourceRepository;
-        this.nrtmFileStore = nrtmFileStore;
+        this.whoisObjectRepository = whoisObjectRepository;
     }
 
-    public List<PublishableSnapshotFile> createSnapshots() throws IOException {
+    public Set<PublishableNrtmFile> createSnapshots() {
```
Comment (MiguelAHM) [inline whois-nrtm4/src/main/java/net/ripe/db/nrtm4/SnapshotFileGenerator.java:70]: this side of code is a bit complicated to understand.
Could we split into modules? which a descriptive name per each module?

File: whois-nrtm4/src/main/java/net/ripe/db/nrtm4/RpslObjectEnqueuer.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+
+
+@Service
+public class RpslObjectEnqueuer {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(RpslObjectEnqueuer.class);
+    private static final int BATCH_SIZE = 100;
+    public static final RpslObjectData POISON_PILL = new RpslObjectData(0, 0, null);
```
[the remark this answers is not in the document]
Comment (MiguelAHM) [reply whois-nrtm4/src/main/java/net/ripe/db/nrtm4/RpslObjectEnqueuer.java:27]: We are using this to do a kind of separation between object when we are writing the json, right?
Maybe OBJECT_DELIMITER or something like this?

File: whois-nrtm4/src/main/java/net/ripe/db/nrtm4/domain/ObjectData.java [hunk 4]
```diff
@@ -0,0 +1,5 @@
+package net.ripe.db.nrtm4.domain;
+
+public record ObjectData(int objectId, int sequenceId) {
```
Comment (MiguelAHM) [inline whois-nrtm4/src/main/java/net/ripe/db/nrtm4/domain/ObjectData.java:3]: LastWhoisObjects?
ObjectData is too ambiguous I think

File: whois-nrtm4/src/main/java/net/ripe/db/nrtm4/util/NrtmFileUtil.java [hunk 5]
```diff
@@ -17,6 +17,11 @@ public static String newFileName(final PublishableNrtmDocument file) {
         return String.format("%s.%d.%s.json", prefix, file.getVersion(), randomHexString());
```
Comment (MiguelAHM) [inline whois-nrtm4/src/main/java/net/ripe/db/nrtm4/util/NrtmFileUtil.java:17]: newFileName visibility could be private, right? we are just using it from newGzFileName

File: whois-nrtm4/src/main/java/net/ripe/db/nrtm4/domain/ObjectData.java (hunk 4 above)
Comment (petchells) [reply whois-nrtm4/src/main/java/net/ripe/db/nrtm4/domain/ObjectData.java:3]: they're not always 'last', that's the thing. it's possible to do updates while this is running so an object might actually be deleted between getting List<ObjectData> and then getting the RPSL (which will then be in 'history'). "WhoisObjectMetaData"? "WhoisObjectState"? if so then what should RpslObjectData become?

File: whois-nrtm4/src/main/java/net/ripe/db/nrtm4/NrtmFileService.java [hunk 6]
```diff
@@ -1,54 +1,126 @@
...
+        try (final FileInputStream fis = NrtmFileUtil.getFileInputStream(path, sessionId, name)) {
+            fis.transferTo(out);
+        }
     }
 
     void syncNrtmFileToFileSystem(final String sessionId, final String name) throws IOException {
         if (!name.startsWith(NrtmDocumentType.DELTA.getFileNamePrefix()) &&
-                !name.startsWith(NrtmDocumentType.SNAPSHOT.getFileNamePrefix())) {
+            !name.startsWith(NrtmDocumentType.SNAPSHOT.getFileNamePrefix())) {
             throw new IllegalArgumentException("Not an NRTM file name: " + name);
```
[the remark this answers is not in the document]
Comment (MiguelAHM) [reply whois-nrtm4/src/main/java/net/ripe/db/nrtm4/NrtmFileService.java:92]: Agree that it is cleaner on this way

File: whois-nrtm4/src/main/java/net/ripe/db/nrtm4/domain/ObjectData.java (hunk 4 above)
Comment (MiguelAHM) [reply whois-nrtm4/src/main/java/net/ripe/db/nrtm4/domain/ObjectData.java:3]: Maybe merge both? In the end is the same just with a new attribute 
Then we can call it WhoisObjectData?

File: whois-nrtm4/src/main/java/net/ripe/db/nrtm4/SnapshotFileGenerator.java [hunk 7]
```diff
@@ -3,92 +3,148 @@
...
+                    Thread.currentThread().interrupt();
+                    throw new RuntimeException(e);
+                }
+            });
+            queueReader.start();
+            queueReaders.add(queueReader);
+        }
+        final Thread queueWriter = new Thread(() -> {
+            try {
+                final int total = state.objectData().size();
```
[the remark this answers is not in the document]
Comment (MiguelAHM) [reply whois-nrtm4/src/main/java/net/ripe/db/nrtm4/SnapshotFileGenerator.java]: I saw that you already changed this. Thank you!

File: whois-nrtm4/src/main/java/net/ripe/db/nrtm4/SnapshotFileGenerator.java (hunk 2 above)
Comment (MiguelAHM) [reply whois-nrtm4/src/main/java/net/ripe/db/nrtm4/SnapshotFileGenerator.java]: Thank you! It is easier to read now

File: whois-nrtm4/src/main/java/net/ripe/db/nrtm4/SnapshotFileGenerator.java [hunk 8]
```diff
@@ -3,92 +3,148 @@
...
+        this.whoisObjectRepository = whoisObjectRepository;
     }
 
-    public List<PublishableSnapshotFile> createSnapshots() throws IOException {
+    public Set<PublishableNrtmFile> createSnapshots() {
+        final Stopwatch stopwatchRoot = Stopwatch.createStarted();
         // Get last version from database.
         final List<NrtmVersionInfo> sourceVersions = nrtmVersionInfoRepository.findLastVersionPerSource();
+        final SnapshotState state = whoisObjectRepository.getSnapshotState();
+        LOGGER.info("NRTM found {} objects in {}", state.objectData().size(), stopwatchRoot);
```
[the remark this answers is not in the document]
Comment (MiguelAHM) [reply whois-nrtm4/src/main/java/net/ripe/db/nrtm4/SnapshotFileGenerator.java]: ah ok! Leave it if you think that it is easier to track the logs, in the end this is the final purpose.

File: whois-nrtm4/src/main/java/net/ripe/db/nrtm4/SnapshotFileGenerator.java [hunk 9]
```diff
@@ -3,96 +3,134 @@
...
-        return snapshotFiles;
+        LOGGER.info("Generation complete {}", stopwatchRoot);
+        return publishedFiles;
     }
 
     public Optional<SnapshotFile> getLastSnapshot(final NrtmSourceModel source) {
         return snapshotFileRepository.getLastSnapshot(source);
     }
 
+    private record RunnableFileGenerator(NrtmFileService nrtmFileService, SnapshotFileRepository snapshotFileRepository,
```
Comment (MiguelAHM) [inline whois-nrtm4/src/main/java/net/ripe/db/nrtm4/SnapshotFileGenerator.java:108]: these parameters should be final, right?

Comment (petchells) [reply whois-nrtm4/src/main/java/net/ripe/db/nrtm4/SnapshotFileGenerator.java:108]: no -- a record is immutable by default so 'final' is implied.

File: whois-nrtm4/src/main/java/net/ripe/db/nrtm4/domain/ObjectData.java (hunk 4 above)
Comment (petchells) [reply whois-nrtm4/src/main/java/net/ripe/db/nrtm4/domain/ObjectData.java]: at the moment ObjectData and RpslObjectData are defined as records. records are immutable, so you can't do 'setRpslObject(...)'; all values must be passed in the constructor. i could use a normal POJO and leave the rpslObject field null until it gets filled but that's hacky -- better to keep the clean separation in definition and in the code, for my money. I'll make them WhoisObjectData and RpslObjectData for now.

File: whois-nrtm4/src/main/java/net/ripe/db/nrtm4/RpslObjectEnqueuer.java [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+    ) {
+        return new RpslObjectQueueRunner(
+            whoisObjectRepository,
+            snapshotState,
+            queueMap,
+            CIString.ciString(whoisSource)
+        );
+    }
+
+    private static class RpslObjectQueueRunner implements Runnable {
```
Comment (MiguelAHM) [inline whois-nrtm4/src/main/java/net/ripe/db/nrtm4/RpslObjectEnqueuer.java:56]: This could be a record, right?

Comment (petchells) [reply whois-nrtm4/src/main/java/net/ripe/db/nrtm4/RpslObjectEnqueuer.java:56]: yes -- i just noticed the notice from intellij. but do you think it's a good idea to have a 'runnable record' class? there are no getters for these constructor vars and i don't think there should be any, so it's not /really/ a record, nor it used like one (i.e. it implements functionality, not a data class like a pojo).

Comment (MiguelAHM) [reply whois-nrtm4/src/main/java/net/ripe/db/nrtm4/RpslObjectEnqueuer.java:56]: so we only will use record for pojo.
ok!

**Review by maggarwal13 — APPROVED** [review]

**Pull request merged**